File: Leetcode/Problems/113. Path Sum II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>>ans;
    void pathSumHelp(TreeNode* root, int sum,vector<int>temp) {
        if(root == NULL)
            return;
        if(sum == root->val && root->left == NULL && root->right == NULL)
        {
            temp.push_back(root->val);
            ans.push_back(temp);
        }
        temp.push_back(root->val);
        pathSumHelp(root->left,sum-root->val,temp);
        pathSumHelp(root->right,sum-root->val,temp);
    }
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<int>temp;
        pathSumHelp(root,sum,temp);
        return ans;
    }
    
};

File: Leetcode/Problems Java/113. Path Sum II.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
     * int val;
     * TreeNode left;
     * TreeNode right;
     * TreeNode() {}
     * TreeNode(int val) { this.val = val; }
     * TreeNode(int val, TreeNode left, TreeNode right) {
         * this.val = val;
         * this.left = left;
         * this.right = right;
     * }
 * }
 */
class Solution {
    List<List<Integer>> ans;

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        ans = new ArrayList<>();
        pathSum(root, targetSum, new ArrayList<>());
        return ans;
    }

    public void pathSum(TreeNode root, int targetSum, List<Integer> res) {
        if (root == null) return;

        List<Integer> newRes = new ArrayList<>();
        newRes.addAll(res);
        newRes.add(root.val);
        
        if (root.left == null && root.right == null && targetSum == root.val) {
            ans.add(newRes);
            return;
        }

        pathSum(root.left, targetSum - root.val, newRes);
        pathSum(root.right, targetSum - root.val, newRes);
    }
}
